File: Go/Day08.go
package main

import AH "./adventhelper"

import (
	"strconv"
	"strings"
)

type Command struct {
	Instruction string
	Value       int
}

func (c *Command) flip() {
	if c.Instruction == "acc" {
		// do nothing
	} else if c.Instruction == "nop" {
		c.Instruction = "jmp"
	} else if c.Instruction == "jmp" {
		c.Instruction = "nop"
	}
}

type Machine struct {
	Accumulator int
	History     []int
	Pointer     int
	Program     []Command
}

func (m Machine) at() (Command) {
	return m.Program[m.Pointer]
}

// exit codes
// 0  - still running
// 1  - halted (pointer out of bounds)
// -1 - entered an infinite loop
// -2 - unrecognised command
func (m *Machine) tick() (exitCode int) {
	if m.Pointer >= len(m.Program) || m.Pointer < 0 {
		return 1 // the pointer is Out of Bounds
	} else if AH.ContainsInt(m.History, m.Pointer) {
		return -1 // we've enetered an infinite loop
	} else { // we are expecting a valid instruction
		ins := m.at().Instruction
		val := m.at().Value
		m.History = append(m.History, m.Pointer)
		if ins == "acc" {
			m.Accumulator += val
			m.Pointer += 1
		} else if ins == "jmp" {
			m.Pointer += val
		} else if ins == "nop" {
			m.Pointer += 1
		} else {
			return -2 // we have a INVALID instruction
		}
		return 0 // Carry on...
	}
}

func (m *Machine) run() (exitCode int) {
  exitCode = 0

	for exitCode == 0 {
		exitCode = m.tick()
	}

	return exitCode	
}

func parseLine(s string) (com Command) {
	parts := strings.Split(s, " ")
	v,_ := strconv.Atoi(parts[1])

	com = Command{Instruction: parts[0], Value: v}
	return com
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input08.txt")
	part1 := 0
	part2 := 0

	prg := make([]Command, 0, len(ss))
	for _,s := range ss {
		q := parseLine(s)
		prg = append(prg, q)
	}

	skynet := Machine{Accumulator: 0,
	                  History:     make([]int, 0, 1000),
	                  Pointer:     0,
	                  Program:     prg}

	skynet.run()
	part1 = skynet.Accumulator
	
	cp := make([]Command, len(prg))
	for i := 0; i < len(prg); i++ {
		copy(cp, prg)
		cp[i].flip()

		tempM := Machine{Accumulator: 0,
		                 History:     make([]int, 0, 1000),
		                 Pointer:     0,
		                 Program:     cp}

		code := tempM.run()

		if code == 1 {
			part2 = tempM.Accumulator
			break
		}
	}

	AH.PrintSoln(8, part1, part2)

	return
}

File: C++/Day08.cpp
#include "AH.h"

namespace Day08
{

	enum instruction { acc = 0, nop, jmp };

	class Command
	{
	public:
		Command(const instruction i, const int64_t v);
		void flip();

		instruction Instr;
		int64_t     Value;
		
	};

	Command::Command(const instruction i, const int64_t v) :
		Instr(i), Value(v) {}

	void Command::flip()
	{
		switch(Instr)
		{
		case instruction::acc:
			break;
		case instruction::nop:
			Instr = instruction::jmp;
			break;
		case instruction::jmp:
			Instr = instruction::nop;
			break;
		}
	}

	class Machine
	{
	public:
		Machine(const std::vector<Command>& cmds);

		Command at() const { return Program[Pointer]; }
		int64_t tick();
		int64_t run();
		int64_t output() { return Accumulator; }

	private:
		int64_t               Accumulator;
		std::vector<uint64_t> History;
		uint64_t              Pointer;
		std::vector<Command>  Program;
		
	};

	Machine::Machine(const std::vector<Command>& cmds) :
		Accumulator(0), Pointer(0), Program(cmds) {}

	int64_t Machine::tick()
	{
		if (Pointer >= Program.size() || Pointer < 0)
			return 1;
		else if (std::find(History.begin(), History.end(), Pointer) != History.end())
			return -1;
		else
		{
			auto ins = Program[Pointer].Instr;
			auto val = Program[Pointer].Value;

			History.push_back(Pointer);
			switch(ins)
			{
			case instruction::acc:
				Accumulator += val;
				Pointer++;
				break;
			case instruction::nop:
				Pointer++;
				break;
			case instruction::jmp:
				Pointer += val;
				break;
			default:
				return -2;		
			}
			return 0;
		}
	}

	int64_t Machine::run()
	{
		int64_t exitCode = 0;
		while (exitCode == 0)
			exitCode = tick();

		return exitCode;
	}

	Command parseLine(const std::string s)
	{
		auto parts = AH::Split(s, ' ');
		int64_t v = std::stoi(parts[1]);
		instruction i = instruction::acc;
		if (parts[0] == "acc")
			i = instruction::acc;
		else if (parts[0] == "jmp")
			i = instruction::jmp;
		else if (parts[0] == "nop")
			i = instruction::nop;

		return Command(i, v);
	}

	int Run(const std::string& filename)
	{
		auto inputLines = AH::ReadTextFile(filename);

		std::vector<Command> cmds;
		std::transform(inputLines.begin(), inputLines.end(),
		               std::back_inserter(cmds),
		               [](std::string s) -> Command { return parseLine(s); });

		Machine m1(cmds);
		m1.run();

		int64_t part2 = 0;
		for (size_t i = 0; i < cmds.size(); ++i)
		{
			if (cmds[i].Instr == instruction::acc)
				continue;

			cmds[i].flip();
			Machine m2(cmds);
			m2.run();

			auto code = m2.run();
			if (code == 1)
			{
				part2 = m2.output();
				break;
			}
			cmds[i].flip();
		}

		AH::PrintSoln(8, m1.output(), part2);

		return 0;
	}

}
